hikari/impl/entity_factory.py

```diff
@@ -1084,23 +1084,19 @@ def deserialize_welcome_screen(self, payload: data_binding.JSONObject) -> guild_
         channels: typing.List[guild_models.WelcomeChannel] = []
 
         for channel_payload in payload["welcome_channels"]:
-            emoji_id = channel_payload["emoji_id"]
-            emoji_name = channel_payload["emoji_name"]
+            raw_emoji_id = channel_payload["emoji_id"]
+            emoji_id = snowflakes.Snowflake(raw_emoji_id) if raw_emoji_id else None
 
-            emoji: typing.Optional[emoji_models.Emoji] = None
-            if emoji_name is not None:
-                if emoji_id is not None:
-                    emoji = emoji_models.CustomEmoji(
-                        id=snowflakes.Snowflake(emoji_id), name=emoji_name, is_animated=None
-                    )
-                else:
-                    emoji = emoji_models.UnicodeEmoji(emoji_name)
+            emoji_name: typing.Union[None, emoji_models.UnicodeEmoji, str]
+            if (emoji_name := channel_payload["emoji_name"]) and not emoji_id:
+                emoji_name = emoji_models.UnicodeEmoji(emoji_name)
```

Comment:
I dont really understand why this logic was changed. This change looks like it should affect (amongst more things) how we serialize the WelcomeChannel not deserialize it

Reply:
This is documented on the class definition; emoji_name isn't always provided when a custom emoji is set for a welcome channel (only emoji_id is guaranteed)

Reply:
For my tests, emoji_name seems to always we provided, with emoji_id being optional:

- `emoji_name` and `emoji_id`: CustomEmoji
- `emoji_name` and not `emoji_id`: UnicodeEmoji
- not `emoji_name` and not `emoji_id`: None

Reply:
it isn't

Reply:
![image](https://user-images.githubusercontent.com/29100934/123985027-6ae35480-d9c5-11eb-87c2-0b31447831db.png)


I am unable to reproduce what you are saying, is there something I am missing?

Reply:
> it's returned as null for custom emojis if it wasn't specified in the patch request

If emoji_name is not specified for a channel where you did specify emoji_id when you PATCHed the resource, emoji_name will be returned as null for that channel

Reply:
Oh true:

![image](https://user-images.githubusercontent.com/29100934/123986828-f3aec000-d9c6-11eb-9980-3342736a06a5.png)


But we already handle this with the current approach, as we allow CustomEmojis to not have a name, due to how weirdly Discord implement this.

Reply:
> EDIT: I just noticed that this should really have a builder behind it

Why? this just matches the pattern seen with stuff like PermissionOverwrites, embeds (but without the fluent interface since these objects still aren't considered mutable), and attachments since it isn't nested at all

Reply:
> But we already handle this with the current approach, as we allow CustomEmojis to not have a name, due to how weirdly Discord implement this.

No we don't? we just ignore the emoji all together if it doesn't have a name set here

Reply:
Well then wouldn't the better approach be to just document `CustomEmoji.name` with the warning you added and just fix it to parse it based on the emoji_id, not the name (like you did here), not split it into two different fields?

Reply:
> Well then wouldn't the better approach be to just document CustomEmoji.name with the warning you added and just fix it to parse it based on the emoji_id, not the name (like you did here), not split it into two different fields everywhere?

Except that doesn't deal with the fact that the interface of CustomEmoji is kinda useless without the use of assertions everywhere it's used since it only guarantees ID as to fit in with two specific edge cases (welcome channel and reaction events) (this is just kinda bad typing, don't exactly want to have to rely on the source of a model to know if it'll work in a specific case cause now any function which takes it will have to document the cases where it'll raise and when using it after getting it from the source you'll have to use assertions based on context which might not even be guaranteed to stay the same), also wouldn't solve the other issues about methods on it returning potentially invalid data when fields are None and also means that CustomEmoji will now have a specific case where CustomEmoji.name.

Reply:
That bug isn't the focus of this PR, that's just something which was picked up since it fell within scope